### PR TITLE
Lazy smp3

### PR DESCRIPTION
**Pull request by mbootsector**
What are we going to do about this?

ELO: 34.41 +-9.9 (95%) LOS: 100.0%
Total: 1094 W: 184 L: 76 D: 834
5000 @ 60+0.05 th 23

ELO: 8.76 +-5.0 (95%) LOS: 100.0%
Total: 5000 W: 735 L: 609 D: 3656
5000 @ 60+0.05 th 7

ELO: -13.52 +-6.5 (95%) LOS: 0.0%
Total: 3625 W: 535 L: 676 D: 2414
5000 @ 15+0.05 th 3

Ongoing VLTC test: http://tests.stockfishchess.org/tests/view/561248460ebc597e4f23e566

### CODE REVIEW

**Comment by VoyagerOne** [comment]
IMO, this should be merge ASAP.  
See how well it does in TCEC in the next round.

**Comment by joergoster** [comment]
Most likely we hit the peak performance of the current YBW code with 3 or maybe 4 threads.
But the gain with more threads is simply too tremendous to ignore, imho.

**Comment by Rocky640** [comment]
There is such a huge discrepancy between different workers, it is hard to follow all these SMP tests.
Would need some excel with workers, specs, and results to better see the trees from the forest.

**Comment by Rocky640** [comment]
Here is JojoM 28 specs. It does very well in the very last test
http://tests.stockfishchess.org/tests/view/561248460ebc597e4f23e566
https://groups.google.com/d/msg/fishcooking/jpzc9c7KlzM/UppIpebkCAAJ

**Comment by jhellis3** [comment]
Given the poor result of the 3 thread test, I would say a longer TC test is needed there.  I would do something like 10k games  60s+0.10.  If that performs OK, then it should be fine to commit IMHO.  Otherwise, more work needs to be done since the vast majority of users is going to fall into the 2-8 thread range (not 16+).

**Comment by burak13** [comment]
We can always address(or not) these low thread problems later on, but for now can't we just keep both smp structures and apply lazy-smp only if threads>8 by UCI parameters? Moreover, we have a great opportunity to observe lazy-smp in the 3rd stage at TCEC. I'm pretty sure lazy-smp's performance won't diminish with longer time control or at least be as good as the current one and will take one of the first two spots easily. According to Louis Zulli's test that posted on TalkChess also gives us a good indication that it doesn't hang or crash.

**Comment by glinscott** [comment]
Very interesting work!!  I don't think we have enough evidence to prove this as better than master yet, as fishtest is having problems with longer games.

Ultimately, I imagine some sort of YBW/lazy hybrid is going to be the best.  So we might need to keep YBW for the top of the tree, and then go lazy once we get closer to the leaves or something like that.

**Comment by mcostalba** [comment]
The hybrid approach seems too complex to me.

What we can do is to create a lazy_smp branch in official repo,  so to
improve it in parallel with master and eventually merge when ready. I can
do it today.

I also suggest to use lazy_smp in next tcec stage, because it seems quite
stable and even in case of 10-20 elo regression we should pass anyhow, so
risk seems low.

**Comment by Stefano80** [comment]
I support Marco's suggestion. I think this is the only way we could have a change of not being crunched by Komodo in the Superfinal.

**Comment by mcostalba** [comment]
Sounds like a good plan.

**Comment by mcostalba** [comment]
We don't want to push the lazy smp stuff in the master yet, because it's so
invasive, but maintaining a separate branch should not be a big deal...

**Comment by mstembera** [comment]
I also support Marco's suggestion to use lazy smp in the next TCEC stage.  We just have to notify Martin not to grab the binary from abrok and provide a custom one.

**Comment by mcostalba** [comment]
Ok, here we are:

https://github.com/official-stockfish/Stockfish/commits/lazy_smp

**Comment by ajithcj** [comment]
It is clear that some version of lazy-smp will go into master eventually. So its an easy decision to have a separate branch to test other improvements to lazy-smp.

But I'm not sure about lazy for next stage TCEC. Problem is that in 6 games vs Komdo in stage-3 if SF-lazy-smp does worse, everyone will blame lazy-smp patch for not scaling at high TC(even though it could be that Komodo is already much better). So I suggest:

1. Use SF-master in stage-3 to see how it performs against Komodo.
2. Work on lazy till superfinals to get a stable working version that works well at all threads
3. Use SF-lazy-smp for superfinals and see if it performs better than stage-3.
4. 6-games and 100 games at stage-3/finals is statistically not good enough. But its the best we can do right now for very long time controls.

Ofcourse it could happen that we use SF-lazy-smp for stage-3 and it does bettern than Komdo. Then we have a clear winner. But I suspect Komodo is atleast as good as SF-lazy-smp as suggested by Bertil's tests at 4mins time controls.

**Comment by joergoster** [comment]
@mcostalba Great! I guess, now we'll see the same high nps towards the endgame just like Komodo. And I think using lazy_smp for the next stage will be a good test for it.

**Comment by mstembera** [comment]
What are the steps I need to perform to get myself to a state where I can submit a pull request for the lazy_smp branch?  I am trying to write a correct fix for the easy move issue.  My git(hub) skills are rather poor.

**Comment by mcostalba** [comment]
@mstembera I have already pushed the lazy_smp branch with your name, you
don't need to do anything more


@aijtchi well, to do not use  it in TCEC 3 for fear of people is not a good
point for me (also because I don't' care people's comments but just tests
results). Going directly to superfinal with a version never tested on TCEC
hardware seems a sub optimal choice to me. We have the possibility to test
lzay_smp before the superfinal on the same superfinal hardware, and with
very low risk. No sense to me do not take advantage of this opportunity.

**Comment by locutus2** [comment]
@mcostalba 
I think also TCEC stage 3 is a great opportunity to test this patch under real conditions with low risk.

**Comment by mstembera** [comment]
@mcostalba 
Thanks Marco.  mbootsector already fixed it in lazy_smp3 except for one line which I commented on.

**Comment by mcostalba** [comment]
@mstembera  and @everybody else

If you have patches against lazy_smp please open a bull request against
lazy_smp branch on official repo:

https://github.com/official-stockfish/Stockfish/tree/lazy_smp

**Comment by mstembera** [comment]
I submitted a test w/ the easy move fix.
http://tests.stockfishchess.org/tests/view/56139e790ebc597e4f23e5ca

**Comment by mbootsector** [comment]
Some quick comments...
I don't think this little patch is anywhere near ready to become the new master. The already created separate branch in the official repo is the best way I think.
No tests confirm yet that it works for many cores and long tc, but it sure looks interesting! Strong hint that the YBWC implementation is not optimally tuned for lots of cores. SMP surely can be improved somehow... I guess the next question is why does lazy SMP work so well in the tests done, but I think I'll leave that to the experts. :)

**Comment by ajithcj** [comment]
@mcostalba, ok sure, I just suggested an alternate option... 
But now that the easymove fix works fine at 3 threads, what is preventing this from coming into master? What additional tests are required?
Also, does fishtest support testing against a branch other than master(I haven't tried it before). This maybe required if someone wants to test an SMP patch against lazy branch.

@mbootsector, congrats!! this is a landmark patch....
I had a quick question:  Why are you not sharing history across threads? Why did you make it thread specific only for lazy SMP?

**Comment by mbootsector** [comment]
@ajithcj I did some TTD tests locally which showed a small improvement when the history table and CM were made thread-local. However, I made all the SMP newbie mistakes and ran 1000 positions to depth 20 only once. Better is to run fewer positions to deeper depths, and run the test at least twice. I don't have more than four cores and the tests take time... Some impatience also played a role here. 
We have not tested a version with shared history and CM on fishtest. There is no proof that thread-local history tables and CM are better.

**Comment by ajithcj** [comment]
@mbootsector, ok thanks for clarifying. I'm not sure how much history will help TTD or NPS. But don't know..

I thought you moved history to threads due to functional issues or to avoid cache thrashing with large thread count. However your test below with original master seems to suggest that cache thrashing is not an issue at high threads:
http://tests.stockfishchess.org/tests/view/560e78790ebc597e4f23e45e

So I suspect you have some more ELO to gain from lazy-smp by using shared history/countermove tables..

**Comment by zamar** [comment]
For now, lazy smp will be maintained in a separate branch.

Please test further lazy smp improvements (and create pull requests) against that branch.

**Comment by mstembera** [comment]
FYI regarding TCEC binary
https://groups.google.com/forum/#!topic/fishcooking/E_Zl22ppPqg
